.github/workflows/ci.yml

```
@@ -0,0 +1,137 @@
+name: CI
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+
+concurrency:
+  group: ${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  # Cabal
+  cabal:
+    name: Cabal / GHC ${{ matrix.ghc }} ${{ matrix.project-variant }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        # TODO: "9.0.2", "9.2.8", "9.4.7", "9.6.3"
+        #       don't forget to update -upper!
+        ghc: ["8.0.2", "8.2.2", "8.4.4", "8.6.5", "8.8.4", "8.10.7", "9.0.2"]
+        project-variant: [""]
+        include:
+          - ghc: 8.0.2
+            project-variant: -lower
+          - ghc: 9.0.2
+            project-variant: -upper
+
+      fail-fast: false
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          show-progress: false
+
+      - name: Setup Haskell
+        uses: haskell-actions/setup@v2
+        id: setup-haskell
+        with:
+          ghc-version: ${{ matrix.ghc }}
+          cabal-update: false
+
+      - name: General Setup
+        run: |
+          cp .ci/cabal.project.local${{ matrix.project-variant}} \
+            cabal.project.local
+          # Print out some information for debugging purposes
+          ghc --version
+          cabal --version
+          cat cabal.project.local
+
+      - name: Setup CI
+        # This generates dist-newstyle/cache/plan.json with hashes for all
+        # dependencies, for cache invalidation.
+        run: |
+          cabal v2-update
+          cabal v2-build all --enable-tests --dry-run
+
+      - name: Restore cached dependencies
+        uses: actions/cache/restore@v3
+        id: cache
+        env:
+          key:
+            ${{ runner.os }}-ghc-${{ matrix.ghc }}-cabal-${{
+            steps.setup-haskell.outputs.cabal-version }}${{
+            matrix.project-variant }}
+        with:
+          path: ${{ steps.setup-haskell.outputs.cabal-store }}
+          key:
+            ${{ env.key }}-${{ hashFiles('cabal.project.local',
+            'dist-newstyle/cache/plan.json') }}
+          restore-keys: ${{ env.key }}-
+
+      - name: Install dependencies
+        run: cabal v2-build all --enable-tests --only-dependencies
+
+      # Cache dependencies already at this point, so that we do not have to
+      # rebuild them should the subsequent steps fail
+      - name: Save cached dependencies
+        uses: actions/cache/save@v3
```

```suggestion
        uses: actions/cache/save@v4
```
and the other locations where `actions/cache` is used; there are deprecation warnings now.